plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Vesting.hs

```diff
@@ -76,103 +89,93 @@ availableFrom (VestingTranche d v) range =
-
-        paidOut = vestingDataPaidOut
-        newAmount = paidOut + amountSpent
-
-        -- Verify that the amount taken out, plus the amount already taken
-        -- out before, does not exceed the threshold that is currently
-        -- allowed
-        amountsValid = newAmount `Value.leq` released
-
-        -- Check that the remaining output is locked by the same validation
-        -- script
-        txnOutputsValid =
-            let remaining = Validation.valueLockedBy ptx (Validation.ownHash ptx) in
-            remaining == (totalAmount d - newAmount)
-
-    in amountsValid && txnOutputsValid
-
-validatorScript :: Vesting -> ValidatorScript
-validatorScript v = Ledger.mkValidatorScript $
-    $$(PlutusTx.compile [|| validatorParam ||])
-        `PlutusTx.applyCode`
-            PlutusTx.liftCode v
-    where validatorParam vd = Scripts.wrapValidator (mkValidator vd)
+        remainingActual  = Validation.valueLockedBy ptx (Validation.ownHash ptx)
+
+        remainingExpected = 
+            remainingFrom vestingTranche1 pendingTxValidRange
+            + remainingFrom vestingTranche2 pendingTxValidRange
+
+    in remainingActual `Value.geq` remainingExpected
+            && Validation.txSignedBy ptx vestingOwner
+
+data Vesting
+instance Scripts.ScriptType Vesting where
+    type instance RedeemerType Vesting = ()
+    type instance DataType Vesting = ()
+
+vestingScript :: VestingParams -> ValidatorScript
+vestingScript = Scripts.validatorScript . scriptInstance
+
+scriptInstance :: VestingParams -> Scripts.ScriptInstance Vesting
+scriptInstance vesting = Scripts.Validator @Vesting
+    ($$(PlutusTx.compile [|| validate ||]) `PlutusTx.applyCode` PlutusTx.liftCode vesting)
+    $$(PlutusTx.compile [|| wrap ||])
+    where
+        wrap = Scripts.wrapValidator @() @()
+            
+contractAddress :: VestingParams -> Ledger.Address
+contractAddress = Scripts.scriptAddress . scriptInstance
+
+vestingContract :: VestingParams -> Contract VestingSchema T.Text ()
+vestingContract vesting = vest <|> retrieve where
+    retrieve = do
+        payment <- endpoint @"retrieve funds"
+        liveness <- retrieveFundsC vesting payment
+        case liveness of
+            Alive -> retrieve
+            Dead  -> pure ()
+    vest = endpoint @"vest funds" >> vestFundsC vesting
+
+payIntoContract :: VestingParams -> Value -> TxOut
+payIntoContract vp value = 
+    Tx.scriptTxOut' 
+        value 
+        (contractAddress vp)
+        (DataScript (PlutusTx.toData ()))
+
+vestFundsC
+    :: ( HasWriteTx s         
+       )
+    => VestingParams
+    -> Contract s T.Text ()
+vestFundsC vesting = do
+    let tx = unbalancedTx [] [payIntoContract vesting (totalAmount vesting)]
+    void $ writeTxSuccess tx
+
+data Liveness = Alive | Dead
+
+retrieveFundsC
+    :: ( HasAwaitSlot s
+       , HasUtxoAt s
+       , HasWriteTx s
+       )
+    => VestingParams
+    -> Value
+    -> Contract s T.Text Liveness
+retrieveFundsC vesting payment = do
```

Comment:
Shouldn't we check if we're allowed to retrieve the money yet and give an error if not?

Reply:
Yeah true that's what I meant by "better error handling" in the commit message. It makes sense to have it though, so I'll add it.

Reply:
The UX is better now, but it's harder to check that the validator script does the right thing because the contract itself won't even attempt to submit the transaction